File: shops/mobilworld.py
from urllib.parse import urlencode

import requests
import collections
from bs4 import BeautifulSoup

from entities import Product
from shops.shop import Shop

collections.Callable = collections.abc.Callable


class MobilWorld(Shop):
    title = "MobilWorld"

    def find(self, query):
        soup = self.send_request(query)

        products = []
        products_html = soup.find_all("div", attrs={"class": "item_block"})
        for product_html in products_html:
            item_info = product_html.find_next("div", attrs={"class": "item_info TYPE_1"})
            item_title = item_info.find_next("div", attrs={"class": "item-title"})
            item_link = item_title.find_next("a", attrs={"class": "dark_link"})
            item_title = item_title.find_next("span")
            price_block = product_html.find_next("div", attrs={"class": "price"})
            price = price_block.find_next("span", attrs={"class": "price_value"})

            product = Product(self.title)
            product.title = item_title.text
            product.price = float(price.text.replace(" ", ""))
            product.link = "https://mobilworld.by" + item_link.attrs.get("href")

            products.append(product)

        return products

    @staticmethod
    def send_request(query):
        url = 'https://mobilworld.by/catalog/'
        data = {
            'q': query,
            's': 'Найти'
        }
        response = requests.get("%s?%s" % (url, urlencode(data)), headers={
            "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 "
                          "Safari/537.36"
        })
        soup = BeautifulSoup(response.text, 'html.parser')

        return soup

File: pkg/shopsparser/shops/mobilworld.go
package shops

import (
	"belby/internal/entities"
	"belby/pkg/request"
	"golang.org/x/net/html"
	"io"
	"net/http"
	"sort"
	"strconv"
	"strings"
)

type MobilWorld struct {
	title string
}

func NewMobilWorldShop() *MobilWorld {
	return &MobilWorld{
		title: "MobilWorld",
	}
}

func (s *MobilWorld) GetTitle() string {
	return s.title
}

func (s *MobilWorld) Find(query string) (entities.Product, error) {
	var products []entities.Product

	response, err := s.makeRequest(query)
	if err != nil {
		return entities.Product{}, err
	}

	products, err = s.parse(query, response)
	if err != nil {
		return entities.Product{}, nil
	}

	sortedProducts, err := s.sort(products)
	if err != nil {
		return entities.Product{}, nil
	}

	if len(sortedProducts) > 0 {
		return sortedProducts[0], nil
	} else {
		return entities.Product{}, nil
	}
}

func (s *MobilWorld) sort(products []entities.Product) ([]entities.Product, error) {
	sortedProducts := products

	sort.Slice(sortedProducts[:], func(i, j int) bool {
		return sortedProducts[i].Price < sortedProducts[j].Price
	})

	return sortedProducts, nil
}

func (s *MobilWorld) parse(query, response string) ([]entities.Product, error) {
	var products []entities.Product

	tokenizer := html.NewTokenizer(strings.NewReader(response))

	divFound := false
	divsCountBeforeClose := 0
	priceFound := false

	product := entities.Product{
		ShopTitle: s.GetTitle(),
	}
	for {
		tn := tokenizer.Next()

		switch {
		case tn == html.ErrorToken:
			return products, nil
		case tn == html.TextToken:
			if divFound {
				if priceFound {
					priceFound = false

					token := tokenizer.Token()

					price, err := strconv.ParseFloat(strings.ReplaceAll(token.Data, " ", ""), 64)
					if err != nil {
						continue
					}
					product.Price = price

					if strings.Contains(product.Title, query) {
						products = append(products, product)
					}
					product = entities.Product{
						ShopTitle: s.GetTitle(),
					}
				}
			}
		case tn == html.StartTagToken:
			token := tokenizer.Token()

			if divFound && token.Data == "div" {
				divsCountBeforeClose++
			}

			if divFound && token.Data == "a" && product.Link == "" {
				linkFound := false
				for _, attribute := range token.Attr {
					if attribute.Key == "class" && attribute.Val == "dark_link" {
						linkFound = true
						break
					}
				}
				if linkFound {
					for _, attribute := range token.Attr {
						if attribute.Key == "href" {
							product.Link = "https://mobilworld.by" + attribute.Val
							break
						}
					}
				}
			}

			if divFound && token.Data == "span" && product.Price == 0 && product.Title != "" {
				for _, attribute := range token.Attr {
					if attribute.Key == "class" && attribute.Val == "price_value" {
						priceFound = true
					}
				}
			}

			for _, attribute := range token.Attr {
				if attribute.Key == "class" && attribute.Val == "catalog_item_wrapp item" {
					divFound = true
					divsCountBeforeClose = 1
				}
			}
		case tn == html.SelfClosingTagToken:
			if divFound {
				token := tokenizer.Token()
				if token.Data == "img" && product.Title == "" {
					for _, attribute := range token.Attr {
						if attribute.Key == "title" {
							product.Title = attribute.Val
							break
						}
					}
				}
			}
		case tn == html.EndTagToken:
			if divFound {
				token := tokenizer.Token()
				if token.Data == "div" {
					divsCountBeforeClose--

					if divsCountBeforeClose == 0 {
						divFound = false
					}
				}
			}
		}
	}
}

func (s *MobilWorld) makeRequest(query string) (string, error) {
	body, err := request.SendRequest(request.Request{
		Method: http.MethodGet,
		URL:    "https://mobilworld.by/catalog/",
		Headers: map[string]string{
			"User-Agent": DefaultUserAgent,
		},
		Query: map[string]string{
			"q":     query,
			"sort":  "PRICE",
			"order": "asc",
			"s":     "Найти",
		},
		Callback: func(body io.ReadCloser) ([]byte, error) {
			return io.ReadAll(body)
		},
	})
	if err != nil {
		return "", err
	}

	return string(body), nil
}
